WirescaleCli/AccessTokenProvider.cs

```diff
@@ -0,0 +1,125 @@
+﻿using System.Diagnostics;
+using System.Net;
+using System.Security.Cryptography;
+using System.Text;
+using System.Web;
+using Auth0.AuthenticationApi;
+using Auth0.AuthenticationApi.Builders;
+using Auth0.AuthenticationApi.Models;
+
+namespace WirescaleCli;
+
+public class AcessTokenProvider
+{
+    private readonly string _auth0ClientId;
+    private readonly string _auth0RedirectUri;
+    private readonly AuthenticationApiClient _authenticationApiClient;
+
+    public AcessTokenProvider(string auth0Domain, string auth0ClientId, string auth0RedirectUri)
+    {
+        _authenticationApiClient = new AuthenticationApiClient(auth0Domain);
+        _auth0ClientId = auth0ClientId;
+        _auth0RedirectUri = auth0RedirectUri;
+    }
+
+    public async Task<string> GetAccessToken()
+    {
+        var codeVerifier = GenerateCodeVerifier();
+        var codeChallenge = GenerateCodeChallenge(codeVerifier);
+
+        var authorizationUrlBuilder = _authenticationApiClient.BuildAuthorizationUrl()
+            .WithAudience("http://localhost:3000")
+            .WithClient(_auth0ClientId)
+            .WithResponseType(AuthorizationResponseType.Code)
+            .WithRedirectUrl(_auth0RedirectUri)
+            .WithValue("code_challenge", codeChallenge)
+            .WithValue("code_challenge_method", "S256")
+            .WithScope("openid");
+
+        var authorizationCode = await CaptureAuthorizationCode(authorizationUrlBuilder);
+
+        var accessTokenResponse = await _authenticationApiClient.GetTokenAsync(new AuthorizationCodePkceTokenRequest
+        {
+            Code = authorizationCode,
+            CodeVerifier = codeVerifier,
+            RedirectUri = _auth0RedirectUri,
+            ClientId = _auth0ClientId,
+        });
+
+        return accessTokenResponse.AccessToken;
+    }
+
+    private async Task<String> CaptureAuthorizationCode(AuthorizationUrlBuilder authorizationUrlBuilder)
+    {
+        var callbackHttpListener = new HttpListener();
+        callbackHttpListener.Prefixes.Add($"{_auth0RedirectUri}/");
+
+        callbackHttpListener.Start();
+
+        OpenBrowser(authorizationUrlBuilder.Build());
+
+        var context = await callbackHttpListener.GetContextAsync();
+        var callbackRequest = context.Request;
+
+
+        var authorizationCode = HttpUtility.ParseQueryString(callbackRequest.Url.Query)["code"];
```

Comment:
Should this check that the received authorization code is not empty?

Reply:
It should -- or rather, we should definitely check if there is an error in the callback url query string. I've [just pushed a change that handles error codes](https://github.com/runesoerensen/wirescale/pull/5/commits/88ee99c2378e76d405685c841aafc567b48a677b) and return relevant error responses (based on the referenced RFC's description of those) in the browser, and prints this error to stderror.

It does not check whether code is null or empty, but based on the RFC I don't believe the redirect will ever happen without a code or an error parameter present. Since the error parameter is now taken care of I think it's safe to assume that the authorization code will always be present - does that sound agreeable? :)